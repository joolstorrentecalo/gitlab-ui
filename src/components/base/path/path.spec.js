import { shallowMount } from '@vue/test-utils';
import items from './examples/data';
import GlPath from './path.vue';

const SELECTED_CLASS_INDIGO = 'gl-path-active-item-indigo';
const SELECTED_CLASS_GREEN = 'gl-path-active-item-green';
const BACKGROUND_COLOR_WHITE = 'white';
const BACKGROUND_COLOR_LIGHT_GRAY = '#f0f0f0';

describe('Path', () => {
  let wrapper;

  const createComponent = (props) => {
    return shallowMount(GlPath, {
      propsData: {
        items,
        ...props,
      },
    });
  };

  const pathNav = () => {
    return wrapper.find('[data-testid="gl-path-nav"]').element;
  };

  const listItems = () => {
    return wrapper.vm.$refs.pathListItems;
  };

  const pathItemAt = (index) => {
    return listItems()[index].children[0];
  };

  const pathItemTextAt = (index) => {
    return pathItemAt(index).textContent;
  };

  const clickItemAt = (index) => {
    pathItemAt(index).click();
  };

  beforeEach(() => {
    wrapper = createComponent();
  });

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  it('matches the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('theme selection', () => {
    describe('default', () => {
      it('applies the indigo theme', () => {
        expect(pathItemAt(0).classList).toContain(SELECTED_CLASS_INDIGO);
      });
    });

    describe('with a theme specified', () => {
      beforeEach(() => {
        wrapper = createComponent({ theme: 'green' });
      });

      it('sets the correct theme', () => {
        expect(pathItemAt(0).classList).toContain(SELECTED_CLASS_GREEN);
      });
    });
  });

  describe('background color selection', () => {
    describe('with no background color specified', () => {
      it('displays the default background color', () => {
        expect(pathNav().style.getPropertyValue('--path-bg-color')).toBe(BACKGROUND_COLOR_WHITE);
      });
    });

    describe('with a background color specified', () => {
      beforeEach(() => {
        wrapper = createComponent({ backgroundColor: BACKGROUND_COLOR_LIGHT_GRAY });
      });

      it('sets the correct background color', () => {
        expect(pathNav().style.getPropertyValue('--path-bg-color')).toBe(
          BACKGROUND_COLOR_LIGHT_GRAY
        );
      });
    });
  });

  describe('renders the list of items', () => {
    it('renders the correct number of items', () => {
      expect(listItems().length).toBe(items.length);
    });

    it('renders the items in the correct order', () => {
      expect(pathItemTextAt(0)).toContain(items[0].title);
      expect(pathItemTextAt(4)).toContain(items[4].title);
      expect(pathItemTextAt(9)).toContain(items[9].title);
    });

    describe('with metrics', () => {
      beforeEach(() => {
        const data = items;
        data[0].metric = '12d';

        wrapper = createComponent({ items: data });
      });

      it('matches the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('renders the inline metric', () => {
        expect(pathItemTextAt(0)).toContain(items[0].title);
        expect(pathItemTextAt(0)).toContain(items[0].metric);
      });
    });

    describe('with icons', () => {
      const iconName = 'home';

      beforeEach(() => {
        const data = items;
        data[0].icon = iconName;

        wrapper = createComponent({ items: data });
      });

      it('matches the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('renders the inline icon', () => {
        const icon = wrapper.find('[data-testid="gl-path-item-icon"]');

        expect(icon.exists()).toBe(true);
        expect(icon.props('name')).toBe(iconName);
      });
    });
  });

  describe('renders the correct selected item', () => {
    describe('with no selected item passed in', () => {
      it('selects the first item', () => {
        expect(pathItemAt(0).classList).toContain(SELECTED_CLASS_INDIGO);
      });
    });

    describe('with a specifically selected item passed in', () => {
      beforeEach(() => {
        const data = items;
        data[3].selected = true;

        wrapper = createComponent({ items: data });
      });

      it('selects the correct item', () => {
        expect(pathItemAt(3).classList).toContain(SELECTED_CLASS_INDIGO);
      });
    });

    describe('with multiple selected items passed in', () => {
      beforeEach(() => {
        const data = items;
        data[3].selected = true;
        data[5].selected = true;

        wrapper = createComponent({ items: data });
      });

      it('selects the first selected option', () => {
        expect(pathItemAt(3).classList).toContain(SELECTED_CLASS_INDIGO);
        expect(pathItemAt(5).classList).not.toContain(SELECTED_CLASS_INDIGO);
      });
    });
  });

  describe('event handling', () => {
    describe('when an item is clicked', () => {
      it('emits the selected event with the correct data', () => {
        clickItemAt(1);
        clickItemAt(4);
        clickItemAt(6);

        expect(wrapper.emittedByOrder()).toEqual([
          { name: 'selected', args: [items[1]] },
          { name: 'selected', args: [items[4]] },
          { name: 'selected', args: [items[6]] },
        ]);
      });
    });
  });
});
