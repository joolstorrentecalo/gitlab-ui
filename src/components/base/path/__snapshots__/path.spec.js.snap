// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Path matches the snapshot 1`] = `
<div
  class="gl-path-nav"
  data-testid="gl-path-nav"
  style="--path-bg-color: white;"
>
  <span
    class="gl-path-fade gl-path-fade-left"
    style="display: none;"
  >
    <button
      aria-label="Scroll left"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-left"
        size="32"
      />
    </button>
  </span>
   
  <ul
    class="gl-path-nav-list"
  >
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-0"
    >
      <button
        class="gl-path-button gl-path-active-item-indigo"
      >
        <!---->
        
        First
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          1d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-1"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Second
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          2d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-2"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Third
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          3d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-3"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Fourth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          4d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-4"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Fifth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          5d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-5"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Sixth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          6d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-6"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Seventh
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          7d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-7"
    >
      <button
        category="tertiary"
        class="gl-path-button"
        disabled="disabled"
      >
        <!---->
        
        Eighth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          8d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-8"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Ninth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          9d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-9"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Tenth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          10d
        </span>
      </button>
       
    </li>
  </ul>
   
  <span
    class="gl-path-fade gl-path-fade-right"
    style="display: none;"
  >
    <button
      aria-label="Scroll right"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-right"
        size="32"
      />
    </button>
  </span>
</div>
`;

exports[`Path renders the list of items with icons matches the snapshot 1`] = `
<div
  class="gl-path-nav"
  data-testid="gl-path-nav"
  style="--path-bg-color: white;"
>
  <span
    class="gl-path-fade gl-path-fade-left"
    style="display: none;"
  >
    <button
      aria-label="Scroll left"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-left"
        size="32"
      />
    </button>
  </span>
   
  <ul
    class="gl-path-nav-list"
  >
    <li
      class="gl-path-nav-list-item"
      id="path-15-item-0"
    >
      <button
        class="gl-path-button gl-path-active-item-indigo"
      >
        <gl-icon-stub
          class="gl-mr-2"
          data-testid="gl-path-item-icon"
          name="home"
          size="16"
        />
        
        First
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          12d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-15-item-1"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Second
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          2d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-15-item-2"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Third
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          3d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-15-item-3"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Fourth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          4d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-15-item-4"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Fifth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          5d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-15-item-5"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Sixth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          6d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-15-item-6"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Seventh
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          7d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-15-item-7"
    >
      <button
        category="tertiary"
        class="gl-path-button"
        disabled="disabled"
      >
        <!---->
        
        Eighth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          8d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-15-item-8"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Ninth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          9d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-15-item-9"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Tenth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          10d
        </span>
      </button>
       
    </li>
  </ul>
   
  <span
    class="gl-path-fade gl-path-fade-right"
    style="display: none;"
  >
    <button
      aria-label="Scroll right"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-right"
        size="32"
      />
    </button>
  </span>
</div>
`;

exports[`Path renders the list of items with metrics matches the snapshot 1`] = `
<div
  class="gl-path-nav"
  data-testid="gl-path-nav"
  style="--path-bg-color: white;"
>
  <span
    class="gl-path-fade gl-path-fade-left"
    style="display: none;"
  >
    <button
      aria-label="Scroll left"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-left"
        size="32"
      />
    </button>
  </span>
   
  <ul
    class="gl-path-nav-list"
  >
    <li
      class="gl-path-nav-list-item"
      id="path-11-item-0"
    >
      <button
        class="gl-path-button gl-path-active-item-indigo"
      >
        <!---->
        
        First
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          12d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-11-item-1"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Second
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          2d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-11-item-2"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Third
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          3d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-11-item-3"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Fourth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          4d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-11-item-4"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Fifth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          5d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-11-item-5"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Sixth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          6d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-11-item-6"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Seventh
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          7d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-11-item-7"
    >
      <button
        category="tertiary"
        class="gl-path-button"
        disabled="disabled"
      >
        <!---->
        
        Eighth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          8d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-11-item-8"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Ninth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          9d
        </span>
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-11-item-9"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Tenth
        <span
          class="gl-font-weight-normal gl-pl-2"
        >
          10d
        </span>
      </button>
       
    </li>
  </ul>
   
  <span
    class="gl-path-fade gl-path-fade-right"
    style="display: none;"
  >
    <button
      aria-label="Scroll right"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-right"
        size="32"
      />
    </button>
  </span>
</div>
`;
